### DISCUSSION TITLE
How to get the user object when the user is logged in and None otherwise?

### DISCUSSION THREAD

**Discussion by gregsifr** [fastapi/fastapi#6565] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from fastapi import Depends, FastAPI
from fastapi.security import OAuth2PasswordBearer
from pydantic import BaseModel

app = FastAPI()

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


class User(BaseModel):
    username: str
    email: Optional[str] = None
    full_name: Optional[str] = None
    disabled: Optional[bool] = None


def fake_decode_token(token):
    return User(
        username=token + "fakedecoded", email="john@example.com", full_name="John Doe"
    )


async def get_current_user(token: str = Depends(oauth2_scheme)):
    user = fake_decode_token(token)
    return user


@app.get("/api_limits")
async def read_users_me(current_user: User = Depends(get_current_user)):
    # current_user to be the User object if the user is logged in and None otherwise 
    if current_user is None:
        return 2
    return 5
```

### Description

How to get the user object if the user is logged in and None otherwise?

Following the [FastAPI documentation for getting the user][1] the `/api_limits` endpoint above returns a 401 when the user is not logged in (in an application with proper jwt signatures).

How to get the `current_user` user object inside the `/api_limits` endpoint in the sample code to be `None` when the user is not logged in?

  [1]: https://fastapi.tiangolo.com/tutorial/security/get-current-user/#create-a-get_current_user-dependency

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

3.8.11

**Comment by jiteshy** [comment]
I think what you are looking for is a custom error handler to override default exception FastAPI returns. As much as suppressing exceptions is discouraged but if it's still needed, you can achieve it with below:

```
from http import HTTPStatus
from fastapi.exceptions import HTTPException
from fastapi.exception_handlers import http_exception_handler

@app.exception_handler(HTTPException)
async def custom_exception_handler(request: Request, exc: HTTPException):
    if exc.status_code == HTTPStatus.UNAUTHORIZED:
        return JSONResponse(
            status_code=HTTPStatus.OK,
            content=None,
        )

    return await http_exception_handler(request, exc)
```

**Comment by YuriiMotov** [comment, marked as answer]
Passing `auto_error=False` to `OAuth2PasswordBearer` will disable raising the exception if token isn't provided. In this case `oauth2_scheme` will return `None` instead of raising exception.

```python
from typing import Optional

from fastapi import Depends, FastAPI
from fastapi.security import OAuth2PasswordBearer
from pydantic import BaseModel

app = FastAPI()

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token", auto_error=False)


class User(BaseModel):
    username: str
    email: Optional[str] = None
    full_name: Optional[str] = None
    disabled: Optional[bool] = None


def fake_decode_token(token):
    return User(
        username=token + "fakedecoded", email="john@example.com", full_name="John Doe"
    )


async def get_current_user(token: str | None = Depends(oauth2_scheme)):
    if token:
        user = fake_decode_token(token)
        return user
    return None


@app.get("/api_limits")
async def read_users_me(current_user: User | None = Depends(get_current_user)):
    if current_user is None:
        return 2
    return 5
```
